crates/hc_demo_cli/src/demo.rs

```diff
@@ -0,0 +1,320 @@
+    let dna_file = gz.finish().unwrap();
+
+    if output == std::path::PathBuf::from("-") {
+        tokio::io::AsyncWriteExt::write_all(&mut tokio::io::stdout(), &dna_file)
+            .await
+            .unwrap();
+    } else {
+        tokio::fs::write(output, &dna_file).await.unwrap();
+    }
+}
+
+async fn run(
+    dna: std::path::PathBuf,
+    inbox: std::path::PathBuf,
+    outbox: std::path::PathBuf,
+    ready: Option<tokio::sync::oneshot::Sender<()>>,
+    rendezvous: Option<holochain::sweettest::DynSweetRendezvous>,
+) {
+    let _ = tokio::fs::create_dir_all(&inbox).await;
+    let _ = tokio::fs::create_dir_all(&outbox).await;
+
+    let dna_gz = if dna.to_string_lossy() == "-" {
+        let mut dna_gz = Vec::new();
+        tokio::io::AsyncReadExt::read_to_end(&mut tokio::io::stdin(), &mut dna_gz)
+            .await
+            .unwrap();
+        dna_gz
+    } else {
+        tokio::fs::read(dna).await.unwrap()
+    };
+
+    let mut dna = Vec::new();
+    std::io::Read::read_to_end(
+        &mut flate2::read::GzDecoder::new(std::io::Cursor::new(dna_gz)),
+        &mut dna,
+    )
+    .unwrap();
+
+    let dna: UnsafeBytes = dna.into();
+    let dna: SerializedBytes = dna.try_into().unwrap();
+    let dna: DnaFile = dna.try_into().unwrap();
+
+    let rendezvous = match rendezvous {
+        Some(rendezvous) => rendezvous,
+        None => {
+            struct PubRendezvous;
+
+            impl holochain::sweettest::SweetRendezvous for PubRendezvous {
+                fn bootstrap_addr(&self) -> &str {
+                    "https://bootstrap.holo.host"
+                }
+
+                fn sig_addr(&self) -> &str {
+                    "wss://holotest.net"
+                }
+            }
+
+            let rendezvous: holochain::sweettest::DynSweetRendezvous = Arc::new(PubRendezvous);
+            rendezvous
+        }
+    };
+
+    let config = holochain::sweettest::SweetConductorConfig::standard();
+
+    let keystore = holochain_keystore::spawn_mem_keystore().await.unwrap();
+
+    let mut conductor = holochain::sweettest::SweetConductor::from_config_rendezvous_keystore(
+        config, rendezvous, keystore,
+    )
+    .await;
+
+    let dna_with_role = holochain::sweettest::DnaWithRole::from(("hc_demo_cli".into(), dna));
+
+    let app = conductor
+        .setup_app("hc_demo_cli", vec![&dna_with_role])
+        .await
+        .unwrap();
+
+    let cell = app.cells().get(0).unwrap().clone();
+    tracing::info!(?cell);
+
+    // PRINT to stdout instead of trace
+    println!("#DNA_HASH#{}#", cell.dna_hash());
+    println!("#AGENT_KEY#{}#", cell.agent_pubkey());
+
+    let i_zome = cell.zome("integrity");
+    tracing::info!(?i_zome);
+    let c_zome = cell.zome("coordinator");
+    tracing::info!(?c_zome);
+
+    let handle = conductor.sweet_handle();
+
+    if let Some(ready) = ready {
+        let _ = ready.send(());
+    }
+
+    loop {
+        let mut dir = tokio::fs::read_dir(&inbox).await.unwrap();
```

Comment:
Does this handle `Ctrl+C`? I'd expect to need a tokio `select!` to listen for the kill signal

Reply:
Luckily, the system handles ctrl-c for us, so the whole process will shut down including any spawned threads or tasks, so we don't have to worry about that. *If* we were trying to run this code within another process and wanted the ability to shut down the demo without shutting down the whole process, we'd have to be more careful to abort any tokio tasks that were spawned.